### ISSUE TITLE
Show current mode in ZSH Vi mode

### ISSUE THREAD

**Issue by stepanzak** [JanDeDobbeleer/oh-my-posh#5438]
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### What would you like to see added?

In ZSH, you can enable Vi mode with `bindkey -v`.
Powerlevel10k has (had) a feature that shows current mode (`normal`/`insert`/`replace` etc.) and immediately updates it on change.
I would love to be able to do the same in omp, since it's really annoying to use the Vi mode whilst not being able to tell whether I'm in `insert` mode or not.

**Comment by gwojan** [comment]
Is there any way this feature could be added to `bash` Vi mode and `pwsh` for that matter? For PowerShell/PSReadLine I currently have a `ViModeChangeHandler` configured and it would be easy to set an environment variable that oh-my-posh could read.

**Comment by JanDeDobbeleer** [comment]
@stepanzak I was playing around with this and adding the following snippet already enables this. Currently thinking about how to enable that from the configuration.

### .zshrc

```zsh
bindkey -v

_omp_redraw-prompt() {
  local precmd
  for precmd in $precmd_functions; do
    $precmd
  done

  zle .reset-prompt
}

function _omp_zle-keymap-select() {
    if [ "${KEYMAP}" = 'vicmd' ]; then
        export POSH_VI_MODE="command"
    else
        export POSH_VI_MODE="insert"
    fi

    _omp_redraw-prompt
}
_omp_create_widget zle-keymap-select _omp_zle-keymap-select

# reset to default mode at the end of line input reading
function _omp_zle-line-finish() {
    export POSH_VI_MODE="insert"
}
_omp_create_widget zle-line-finish _omp_zle-line-finish

# Fix a bug when you C-c in CMD mode, you'd be prompted with CMD mode indicator
# while in fact you would be in INS mode.
# Fixed by catching SIGINT (C-c), set mode to INS and repropagate the SIGINT,
# so if anything else depends on it, we will not break it.
TRAPINT() {
    export POSH_VI_MODE="insert"
    return $(( 128 + $1 ))
}
```

And use the following code block to display this:

### mytheme.omp.json

```json
{
  "type": "text",
  "style": "plain",
  "template": "{{ if .Env.POSH_VI_MODE }}({{ .Env.POSH_VI_MODE }}) {{ end }}"
}
```

**Comment by NovaViper** [comment]
Hey, would it be possible for you to also make it where the prompt symbol can change based on what mode we're in? Possibly integrate with the [zsh-vi-mode plugin](https://github.com/jeffreytse/zsh-vi-mode) as an additional nicety!

**Comment by JanDeDobbeleer** [comment]
> integrate with the [zsh-vi-mode plugin](https://github.com/jeffreytse/zsh-vi-mode) as an additional nicety!

@NovaViper that's not something we do, but changing the icon is rather straightforward with the use of a template

**Comment by NovaViper** [comment]
@JanDeDobbeleer Ah ok! I was wondering because I can't consistently get the variable to display in the prompt. I can get it to show `insert` but oddly can't get `command` to appear when I hit the escape key to enter normal mode; even with the zsh-vi-mode plugin removed. What happens is when I first load up the terminal, the variable is empty (not taking into account the current mode when I start up is in `viins`). And when I hit enter, it shows `insert`; but `command` under any circumstance does not show at all.

This is my zshrc (I'm using NixOS with Home-Manager to setup zsh), https://gist.github.com/NovaViper/6eeb096936b0c52290632986a3a18d38

**Comment by Frestein** [comment]
> Hey, would it be possible for you to also make it where the prompt symbol can change based on what mode we're in? Possibly integrate with the [zsh-vi-mode plugin](https://github.com/jeffreytse/zsh-vi-mode) as an additional nicety!

My solution:
![image](https://github.com/user-attachments/assets/aa681d78-a09f-4321-8260-895693bc26af)
![image](https://github.com/user-attachments/assets/c066ef49-fc61-4500-b5cb-7d047c3b2fa6)
![image](https://github.com/user-attachments/assets/f04a295d-2310-498c-a08b-989d3d76c6ad)
![image](https://github.com/user-attachments/assets/a837c7bf-5652-45f2-97e7-3aa8a0db0977)
![image](https://github.com/user-attachments/assets/af00e603-a23c-4191-babc-2e77fa885fe6)

```sh
# OMP zsh-vi-mode integration
_omp_redraw-prompt() {
  local precmd
  for precmd in "${precmd_functions[@]}"; do
    "$precmd"
  done

  zle .reset-prompt
}

export POSH_VI_MODE="I"

function zvm_after_select_vi_mode() {
  case $ZVM_MODE in
  $ZVM_MODE_NORMAL)
    POSH_VI_MODE="N"
    ;;
  $ZVM_MODE_INSERT)
    POSH_VI_MODE="I"
    ;;
  $ZVM_MODE_VISUAL)
    POSH_VI_MODE="V"
    ;;
  $ZVM_MODE_VISUAL_LINE)
    POSH_VI_MODE="V-L"
    ;;
  $ZVM_MODE_REPLACE)
    POSH_VI_MODE="R"
    ;;
  esac
  _omp_redraw-prompt
}
```
```toml
[[blocks.segments]]
leading_diamond = ''
trailing_diamond = ' '
template = '{{ if .Env.POSH_VI_MODE }}{{ .Env.POSH_VI_MODE }}{{ end }}'
foreground = 'red'
background = '#3C3836'
type = 'text'
style = 'diamond'
foreground_templates = [
  "{{if contains \"N\" .Env.POSH_VI_MODE}}blue{{end}}",
  "{{if contains \"I\" .Env.POSH_VI_MODE}}magenta{{end}}",
  "{{if contains \"V\" .Env.POSH_VI_MODE}}cyan{{end}}",
  "{{if contains \"V-L\" .Env.POSH_VI_MODE}}cyan{{end}}",
  "{{if contains \"R\" .Env.POSH_VI_MODE}}yellow{{end}}",
]
```

**Comment by NovaViper** [comment]
@Frestein Thanks! I took your example and further modified it to allow for the prompt character to change!
```
# OMP zsh-vi-mode integration
_omp_redraw-prompt() {
  local precmd
  for precmd in "${precmd_functions[@]}"; do
    "$precmd"
  done
  zle && zle reset-prompt
}

export POSH_VI_MODE="$(printf '\u276F')"

function zvm_after_select_vi_mode() {
  case $ZVM_MODE in
  $ZVM_MODE_NORMAL)
    POSH_VI_MODE="$(printf '\u276E')"
    ;;
  $ZVM_MODE_INSERT)
    POSH_VI_MODE="$(printf '\u276F')"
    ;;
  $ZVM_MODE_VISUAL)
    POSH_VI_MODE="$(printf '\u2164')"
    ;;
  $ZVM_MODE_VISUAL_LINE)
    POSH_VI_MODE="$(printf '\u2164')"
    ;;
  $ZVM_MODE_REPLACE)
    POSH_VI_MODE="$(printf '\u25B6')"
    ;;
  esac
  _omp_redraw-prompt
}
```

This is translated from my json file that my NixOS configuration generates, may not work exactly as intended.
```
[var]
PromptChar = "❯"

[[blocks]]
type = "prompt"
newline = "true"
alignment = "left"
    [[blocks.segments]]
    background = "transparent"
    foreground_templates = [
        "{{if gt .Code 0}}1{{end}}",
        "{{if eq .Code 0}}10{{end}}"
    ]
    style = "plain"
    template = "{{ if .Env.POSH_VI_MODE }}{{ .Env.POSH_VI_MODE }}{{ else }}{{.Var.PromptChar}}{{ end }}"
    type = "text"
```
Quick Edit: Here's a reference of the commit for anyone who needs it for their own NixOS setup! https://github.com/NovaViper/NixConfig/commit/13e0d01a8d7e6d150a0174fc9b55dc40c73daaa7

**Comment by stepanzak** [comment]
Hi, when I add this code to my config, it creates a loop printing `stty: 'standard input': Inappropriate ioctl for device` forever once I enter normal mode using escape.
![image](https://github.com/user-attachments/assets/086bc164-dc52-4ece-84e8-5229fae5f8e2)
I really tried to find the cause myself but it looks like I don't have enough skill to find it. Does it still work for you? It doesn't for me even when I unset ZDOTDIR and run `zsh -df`, which sould run default zsh without any other config that might have caused the problem.
(edit: this is a reaction to @JanDeDobbeleer 's comment, not any of the others)

**Comment by leojweda** [comment]
@stepanzak Were you able to find a solution to that problem?

I've narrowed it down to the bit below, but the prompt doesn't update without it:

```
  local precmd
  for precmd in "${precmd_functions[@]}"; do
    "$precmd"
  done
```

**Comment by stepanzak** [comment]
@leojweda I wasn't, sadly.

**Comment by Chris-1101** [comment]
@stepanzak @leojweda (CC @JanDeDobbeleer)

TLDR: The loop seems to be caused by having the top-level config option `enable_cursor_positioning` enabled.

---

If you look at the output of `oh-my-posh init zsh` command without `eval`ing it, you can inspect the contents of the `_omp_precmd` function, which is the last of the "precmd" functions run in the `_omp_redraw-prompt` loop. This function in turn calls `_omp_set_cursor_position`.

```zsh
function _omp_set_cursor_position
{
    # not supported in Midnight Commander
    # see https://github.com/JanDeDobbeleer/oh-my-posh/issues/3415
    if [[ $_omp_cursor_positioning == 0 ]] || [[ -v MC_SID ]]; then
        return
    fi

    local oldstty=$(stty -g)
    stty raw -echo min 0

    local pos
    echo -en '\033[6n' >/dev/tty
    read -r -d R pos
    pos=${pos:2} # strip off the esc-[
    local parts=(${(s:;:)pos})

    stty $oldstty

    export POSH_CURSOR_LINE=${parts[1]}
    export POSH_CURSOR_COLUMN=${parts[2]}
}
```
This runs `stty` three times, which is the exact amount of errors printed per loop. I haven't investigated why this failure causes an endless loop, but that error is usually associated with `stty` being run in a non-interactive shell. Disabling `enable_cursor_positioning` fixed it for me.

However, even with it "working", my prompt jumps up one line every time I change modes, and my right prompt ends up stuck to the left one. Perhaps this is due to me having a right-aligned `prompt` instead of an `rprompt` + a second line prompt.

If I figure out a fix, I'll follow up. But for now, this might help those with a simpler prompt 🙃

**Comment by bad-noodles** [comment]
This is my current solution with the default ZSH vi mode

```zsh
bindkey -v

redraw-prompt() {
  _omp_precmd
  zle .reset-prompt
}

vimode-cmd() {
  export VI_MODE="NORMAL"
  redraw-prompt
}

vimode-insert() {
  export VI_MODE="INSERT"
  redraw-prompt
}

vimode-visual() {
  export VI_MODE="VISUAL"
  redraw-prompt
}

vimode-visual-line() {
  export VI_MODE="V-LINE"
  redraw-prompt
}

vimode-replace() {
  export VI_MODE="REPLACE"
  redraw-prompt
}


# CMD/Normal mode
_omp_create_widget vi-cmd-mode vimode-cmd
_omp_create_widget deactivate-region vimode-cmd

# Insert mode
_omp_create_widget vi-insert vimode-insert
_omp_create_widget vi-insert-bol vimode-insert
_omp_create_widget vi-add-eol vimode-insert
_omp_create_widget vi-add-next vimode-insert
_omp_create_widget vi-change vimode-insert
_omp_create_widget vi-change-eol vimode-insert
_omp_create_widget vi-change-whole-line vimode-insert
_omp_create_widget vi-open-line-above vimode-insert
_omp_create_widget vi-open-line-below vimode-insert


# Replace mode
_omp_create_widget vi-replace vimode-replace
_omp_create_widget vi-replace-chars vimode-replace

# Visual mode
_omp_create_widget visual-mode vimode-visual
_omp_create_widget visual-line-mode vimode-visual-line


# reset to default mode at the end of line input reading
line-finish() {
    export VI_MODE="INSERT"
}
_omp_create_widget zle-line-finish line-finish

# Fix a bug when you C-c in CMD mode, you'd be prompted with CMD mode indicator
# while in fact you would be in INS mode.
# Fixed by catching SIGINT (C-c), set mode to INS and repropagate the SIGINT,
# so if anything else depends on it, we will not break it.
TRAPINT() {
  vimode-insert
  return $(( 128 + $1 ))
}

export VI_MODE="INSERT"
```

And this is how I am using it in a segment:

```json
        {
          "properties": {
            "cache_duration": "none"
          },
          "leading_diamond": "\ue0b6",
          "template": "{{ if .Env.VI_MODE }} {{ .Env.VI_MODE }} {{ end }}",
          "background": "#5898ff",
          "background_templates": [
            "{{if eq .Env.VI_MODE \"NORMAL\"}}#5898ff{{end}}",
            "{{if eq .Env.VI_MODE \"INSERT\"}}#ff57a8{{end}}",
            "{{if eq .Env.VI_MODE \"VISUAL\"}}#b879ff{{end}}",
            "{{if eq .Env.VI_MODE \"V-LINE\"}}#b879ff{{end}}",
            "{{if eq .Env.VI_MODE \"REPLACE\"}}#ff2185{{end}}"
          ],
          "type": "text",
          "style": "diamond"
        }
 ```

**Comment by bad-noodles** [comment]
> [@stepanzak](https://github.com/stepanzak) Were you able to find a solution to that problem?
> 
> I've narrowed it down to the bit below, but the prompt doesn't update without it:
> 
> ```
>   local precmd
>   for precmd in "${precmd_functions[@]}"; do
>     "$precmd"
>   done
> ```

I don't believe you need to call every function in `precmd_functions` as you don't want to update the values provided by all of them.
I am calling only `_omp_precmd` and seems to be enough.

**Comment by smitropoulos-arista** [comment]
Guys I am getting

```bash
/home/smitropoulos/.zshrc:126: command not found: _omp_create_widget
```

I suppose I should be including something?

**Comment by bad-noodles** [comment]
> Guys I am getting
> 
> /home/smitropoulos/.zshrc:126: command not found: _omp_create_widget
> I suppose I should be including something?

Did you install oh-my-posh and enable it in you `.zshrc` with `eval "$(oh-my-posh init zsh)"`?

**Comment by smitropoulos-arista** [comment]
Yes, this is the relevant section

```bash

eval "$(oh-my-posh init zsh --config ~/.sm_ohmyposh.json)"

bindkey -v

redraw-prompt() {
  _omp_precmd
  zle .reset-prompt
}

vimode-cmd() {
  export VI_MODE="NORMAL"
  redraw-prompt
}

vimode-insert() {
  export VI_MODE="INSERT"
  redraw-prompt
}

vimode-visual() {
  export VI_MODE="VISUAL"
  redraw-prompt
}

vimode-visual-line() {
  export VI_MODE="V-LINE"
  redraw-prompt
}

vimode-replace() {
  export VI_MODE="REPLACE"
  redraw-prompt
}


# CMD/Normal mode
_omp_create_widget vi-cmd-mode vimode-cmd
_omp_create_widget deactivate-region vimode-cmd

# Insert mode
_omp_create_widget vi-insert vimode-insert
_omp_create_widget vi-insert-bol vimode-insert
_omp_create_widget vi-add-eol vimode-insert
_omp_create_widget vi-add-next vimode-insert
_omp_create_widget vi-change vimode-insert
_omp_create_widget vi-change-eol vimode-insert
_omp_create_widget vi-change-whole-line vimode-insert
_omp_create_widget vi-open-line-above vimode-insert
_omp_create_widget vi-open-line-below vimode-insert


# Replace mode
_omp_create_widget vi-replace vimode-replace
_omp_create_widget vi-replace-chars vimode-replace

# Visual mode
_omp_create_widget visual-mode vimode-visual
_omp_create_widget visual-line-mode vimode-visual-line


# reset to default mode at the end of line input reading
line-finish() {
    export VI_MODE="INSERT"
}
_omp_create_widget zle-line-finish line-finish

# Fix a bug when you C-c in CMD mode, you'd be prompted with CMD mode indicator
# while in fact you would be in INS mode.
# Fixed by catching SIGINT (C-c), set mode to INS and repropagate the SIGINT,
# so if anything else depends on it, we will not break it.
TRAPINT() {
  vimode-insert
  return $(( 128 + $1 ))
}

export VI_MODE="INSERT"
```

Does it matter how you install oh-my-posh for this way to work?

**Comment by sageil** [comment]
Here is what I came up reverse the prompt direction when in vi mode using [zsh-vi-mode plugin](https://github.com/jeffreytse/zsh-vi-mode)
```bash
export SHELL_PROMPT='❯'
_omp_redraw-prompt() {
  local precmd
  for precmd in "${precmd_functions[@]}"; do
    "$precmd"
  done

  zle .reset-prompt
}
function zle-line-init zle-keymap-select {
    case ${KEYMAP} in
       (vicmd)       SHELL_PROMPT='❮' ;;
       (main|viins)  SHELL_PROMPT='❯' ;;
       (*)           SHELL_PROMPT='❮' ;;
    esac
    _omp_redraw-prompt
}


zle -N zle-line-init
zle -N zle-keymap-select

```

```toml

template = '{{ .Env.SHELL_PROMPT }}'
```

**Comment by saboco** [comment]
@bad-noodles your solution works great, thank you!
There is a bug, though, and I haven’t been able to find a solution. If you type a command, then go to normal mode and hit Enter so the command is executed and a new prompt is drawn, the mode stays in normal even though we are now in insert mode.
